src/main/java/com/bootcamp/demo/service/UserService.java

```diff
@@ -0,0 +1,151 @@
+                    .document(userId)
+                    .get().get();
+
+            if (!user.exists()) {
+                LOGGER.warn("DELETE USER BY ID - User does not exit! UserId = {}", userId);
+            } else {
+
+                ApiFuture<WriteResult> collectionApiFuture = db.collection(COLLECTION_USERS_PATH).
+                        document(userId).delete();
+
+                String updateTime = collectionApiFuture.get().getUpdateTime().toDate().toString();
+
+                LOGGER.info("DELETE USER BY ID - service function completed successfully. Update time: {}", updateTime);
+            }
+
+
+        } catch (ExecutionException | InterruptedException e) {
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public User getUserById(final String userId) {
+        LOGGER.info(String.format("FIND USER BY ID - service function invoked. UserId = %s", userId));
+
+        try {
+            DocumentSnapshot user = db.collection(COLLECTION_USERS_PATH)
+                    .document(userId)
+                    .get()
+                    .get();
+
+            if (!user.exists()) {
+                LOGGER.warn("FIND USER BY ID - user does not exist! userId = {}", userId);
+                throw new ItemNotFoundException("User does not exist! userID = " + userId);
+            }
+
+            return user.toObject(User.class);
+
+        } catch (ExecutionException | InterruptedException e) {
+            LOGGER.error("GET USER BY ID - service function. Error message: {}", e.getMessage(), e);
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public User getUserByEmail(final String email) {
+        LOGGER.info(String.format("GET USER BY EMAIL - service function invoked. UserId = %s", email));
+
+        try {
+            List<QueryDocumentSnapshot> filtered = db.collection(COLLECTION_USERS_PATH)
+                    .whereEqualTo("email", email)
+                    .get().get()
+                    .getDocuments();
+
+            if (filtered.isEmpty()) {
+                LOGGER.warn("GET USER BY EMAIL - user does not exist! Email = {}", email);
+                throw new ItemNotFoundException("User does not exist! User email = " + email);
+            }
+
+            DocumentSnapshot user = filtered.get(0);
+
+            return user.toObject(User.class);
+
+        } catch (ExecutionException | InterruptedException e) {
+            LOGGER.error("GET USER BY EMAIL - service function. Error message: {}", e.getMessage(), e);
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public Set<User> getUsers() {
+        LOGGER.info("GET USERS - service function invoked");
+
+        try {
+            List<QueryDocumentSnapshot> users = db.collection(COLLECTION_USERS_PATH)
+                    .get()
+                    .get()
+                    .getDocuments();
+
+            return StreamSupport.stream(users.spliterator(), false)
```

Comment:
You can simplify this by using `.stream()` on the documents' list directly. Also, you would not need the variable and could have:
```
return db.collection(COLLECTION_USERS_PATH)
        .get()
        .get()
        .getDocuments()
        .stream()
        .map(qds -> qds.toObject(User.class))
        .collect(Collectors.toSet());
```